controllers/funding/materials.js

```diff
@@ -10,60 +10,78 @@ const Raven = require('raven');
+    const isValidAction = validActions.indexOf(action) !== -1;
 
-    if (isValidAction) {
-        const maxQuantity = itemToBeAdded.maximum;
-        const notAllowedWithItemId = itemToBeAdded.notAllowedWithItem;
+    // create the basket if empty
+    if (!req.session[sessionOrderKey]) {
+        req.session[sessionOrderKey] = [];
+    }
 
-        const allCurrentOrders = get(req.session, [orderKey], {});
+    // Reset the blocker flag
+    delete req.session[sessionBlockedItemKey];
+
+    if (isValidAction) {
+        let existingProduct = req.session[sessionOrderKey].find(order => order.productId === productId);
 
         // How many of the current item do they have?
-        const currentItemQuantity = get(req.session, [orderKey, code, 'quantity'], 0);
+        const currentItemQuantity = existingProduct ? existingProduct.quantity : 0;
 
         // Check if their current orders contain a blocker
-        const hasBlockerItem = some(allCurrentOrders, order => {
-            return notAllowedWithItemId && order.id === notAllowedWithItemId && order.quantity > 0;
+        // note that this only works in one direction:
+        // eg. it only checks if the product you're adding has constraints
+        // eg. item A is blocked with item B (but you can add item B first, then add A)
+        // solution is to make item A block item B and item B block item A in the CMS
+        // or track the blocked items here and check both ends.
+        const hasBlockerItem = some(req.session[sessionOrderKey], order => {
+            if (!notAllowedWith) {
+                return;
+            }
+            const itemIsBlocked = notAllowedWith.indexOf(order.materialId) !== -1;
+            return itemIsBlocked && order.quantity > 0;
         });
 
-        // Store the product name
-        set(req.session, [orderKey, code, 'name'], itemToBeAdded.name.en);
-
         const noSpaceLeft = currentItemQuantity === maxQuantity;
 
-        // Reset the blocker flag
-        set(req.session, [orderKey, 'itemBlocked'], false);
-
-        if (action === 'increase') {
-            if (!noSpaceLeft && !hasBlockerItem) {
-                set(req.session, [orderKey, code, 'id'], id);
-                set(req.session, [orderKey, code, 'quantity'], currentItemQuantity + 1);
-            } else {
-                // Alert the user that they're blocked from adding this item
-                set(req.session, [orderKey, 'itemBlocked'], true);
-            }
-        } else if (currentItemQuantity > 1 && action === 'decrease') {
-            set(req.session, [orderKey, code, 'id'], id);
-            set(req.session, [orderKey, code, 'quantity'], currentItemQuantity - 1);
-        } else if (action === 'remove' || (action === 'decrease' && currentItemQuantity === 1)) {
-            set(req.session, [orderKey, code, 'quantity'], 0);
+        if (action === 'increase' && (hasBlockerItem || noSpaceLeft)) {
```

Comment:
`!` conditions always read a little weirdly to me and the `action === 'increase' && (hasBlockerItem || noSpaceLeft)` check might need a name. Does this feel any simpler:

```js
const shouldBlock = action === 'increase' && (hasBlockerItem || noSpaceLeft);
if (existingProduct) {
    …
} else if (shouldBlock) {
    …
} else {
    …
}
```

Reply:
The `shouldBlock` name works, but the logic is in this sequence for a reason – the `if (existingProduct)` check happens second because by that point we've ruled out that it's not in invalid increase (eg. they're at maximum, or blocked). If we put that check first we'd have to have nested/repeated checks for quantity/blocker status again.